serde_derive/src/internals/check.rs

```diff
@@ -408,6 +409,43 @@ fn check_transparent(cx: &Ctxt, cont: &mut Container, derive: Derive) {
     }
 }
 
+/// Externally tagged/untagged enum variants must have string names.
+fn check_non_string_renames(cx: &Ctxt, cont: &mut Container) {
+    let details = match &cont.data {
+        Data::Enum(_, _) => match cont.attrs.tag() {
+            TagType::Adjacent { .. } | TagType::Internal { .. } => return,
+            TagType::External => "externally tagged enums",
+            TagType::None => "untagged enums",
+        },
+        Data::Struct(_, _) => return,
+    };
+
+    if let Data::Enum(_, variants) = &cont.data {
```

Comment:
This `if` is always true.

It would be better to capture the required `variants` in the previous match above.